test/async-test.spec.ts

```diff
@@ -0,0 +1,172 @@
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      expect(finished).toBe(true);
+      done();
+    }, () => {
+      done.fail('async zone called failCallback unexpectedly');
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      setTimeout(() => {
+        finished = true;
+      });
+    });
+  });
+
+  it('should call finish after microtasks are done', (done) => {
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      expect(finished).toBe(true);
+      done();
+    }, () => {
+      done.fail('async zone called failCallback unexpectedly');
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      Promise.resolve().then(() => {
+        finished = true;
+      });
+    });
+  });
+
+  it('should call finish after an event task is done', (done) => {
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      expect(finished).toBe(true);
+      done();
+    }, () => {
+      done.fail('async zone called failCallback unexpectedly');
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      var div = document.createElement('div');
+
+      div.addEventListener('click', () => {
+        finished = true;
+      });
+
+      div.click();
+    });
+  });
+
+  it('should fail if setInterval is used', (done) => {
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      done.fail('expected failCallback to be called');
+    }, (err) => {
+      expect(err).toEqual('Cannot use setInterval from within an async zone test.');
+      done();
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      setInterval(() => {
+      }, 100);
+    });
+  });
+
+  it('should fail if an error is thrown', (done) => {
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      done.fail('expected failCallback to be called');
+    }, (err) => {
+      expect(err).toEqual('my error');
+      done();
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      setTimeout(() => {
+        throw new Error('my error');
+      }, 10);
+    });
+  });
+
+  it('should fail if a promise rejection is unhandled', (done) => {
+    var finished = false;
+
+    var testZoneSpec = new AsyncTestZoneSpec(() => {
+      done.fail('expected failCallback to be called');
+    }, (err) => {
+      expect(err).toEqual('Uncaught (in promise): my reason');
+      done();
+    }, 'name');
+
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      // Promise rejection does not cause onHandleError to trigger
+      // until the next Macro task, so that there's a chance
+      // for handlers to be added.
+      Promise.reject('my reason');
+    });
+
+  });
+});
```

Comment:
Done! Added some. Also, that uncovered some bugs with the event task searching, so I fixed it up. Basically, we never want to wait on event tasks.